uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/controllers/leave-request-popup-controller_test.js

```diff
@@ -0,0 +1,591 @@
+            expect($ctrl.uiOptions.selectedToType).not.toBeDefined();
+            expect($ctrl.balance.closing).toEqual(0);
+            expect($ctrl.balance.change.amount).toEqual(0);
+          });
+        });
+      });
+
+      describe('calendar', function () {
+
+        it('no dates are selected by default', function () {
+          expect($ctrl.uiOptions.fromDate).not.toBeDefined();
+          expect($ctrl.uiOptions.toDate).not.toBeDefined();
+        });
+
+        describe('from date', function () {
+
+          beforeEach(function () {
+            $ctrl.onDateChange(new Date(), true);
+            $scope.$digest();
+          });
+
+          it('can change date', function () {
+            expect(moment($ctrl.leaveRequest.from_date, serverDateFormat, true).isValid()).toBe(true);
+          });
+
+        });
+
+        describe('to date', function () {
+
+          beforeEach(function () {
+            $ctrl.onDateChange(new Date(), false);
+            $scope.$digest();
+          });
+
+          it('can change date', function () {
+            expect(moment($ctrl.leaveRequest.to_date, serverDateFormat, true).isValid()).toBe(true);
+          });
+
+        });
+      });
+
+      describe('day types', function () {
+
+        it('by default no selection', function () {
+          expect($ctrl.uiOptions.selectedFromType).not.toBeDefined();
+          expect($ctrl.uiOptions.selectedToType).not.toBeDefined();
+        });
+
+        describe('from', function () {
+
+          beforeEach(function () {
+            spyOn($ctrl, 'filterLeaveRequestDayTypes').and.callThrough();
+            $ctrl.onDateChange(new Date(), true);
+            $scope.$digest();
+          });
+
+          it('by default selects first day type', function () {
+            expect($ctrl.uiOptions.selectedFromType).toEqual($ctrl.leaveRequestFromDayTypes[0]);
+          });
+
+          it('will filter based on current date', function () {
+            expect($ctrl.filterLeaveRequestDayTypes).toHaveBeenCalledWith(jasmine.any(Date), jasmine.any(Boolean));
+          });
+
+          describe('is public holiday', function () {
+
+            beforeEach(function () {
+              $ctrl.uiOptions.fromDate = new Date('01/01/2016');
+              $ctrl.filterLeaveRequestDayTypes($ctrl.uiOptions.fromDate, true);
+              $scope.$digest();
+            });
+
+            it('should', function () {
+              expect($ctrl.leaveRequestFromDayTypes.length).toEqual(1);
+            });
+          });
+
+        });
+
+        describe('to', function () {
+
+          beforeEach(function () {
+            $ctrl.onDateChange(new Date(), false);
+            $scope.$digest();
+          });
+
+          it('by default selects first day type', function () {
+            expect($ctrl.uiOptions.selectedToType).toEqual($ctrl.leaveRequestToDayTypes[0]);
+          });
+
+          describe('change selection', function () {
```

Comment:
This is confusing, you have "from" and "to" day types, but for each of them you are doing different tests? Like for "from" you are doing that "public holidays" test, but for "to" you are doing this "change selection" test?